File: src/year2021/month1/no49/GroupAnagrams.java
package year2021.month1.no49;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.IntStream;

public class GroupAnagrams {
    public static void main(String[] args) {
        String[] strs1 = new String[]{"eat", "tea", "tan", "ate", "nat", "bat"};
        disp(groupAnagrams(strs1));
    }

    public static List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> result = new ArrayList<>();
        if (strs == null || strs.length == 0) {
            return result;
        }
        int[][] alphaSet = new int[strs.length][26];
        int[] sizes = new int[strs.length];
        for (int i = 0; i < strs.length; i++) {
            String str = strs[i];
            boolean isNewGroup = true;
            for (int j = 0; j < result.size(); j++) {
                if (str.length() != sizes[j]) {
                    continue;
                }
                if (isAnagrams(alphaSet, str, j)) {
                    result.get(j).add(str);
                    isNewGroup = false;
                    break;
                }
            }
            if (isNewGroup) {
                int index = result.size();
                List<String> list = new LinkedList<>();
                list.add(str);
                result.add(list);
                sizes[index] = str.length();
                for (char c : str.toCharArray()) {
                    alphaSet[index][c - 'a']++;
                }
            }
        }
        return result;
    }

    private static boolean isAnagrams(int[][] alphaSet, String str, int index) {
        int[] set = Arrays.copyOf(alphaSet[index], alphaSet[index].length);
        for (char c : str.toCharArray()) {
            set[c - 'a']--;
        }
        return IntStream.of(set).allMatch(num -> num == 0);
    }

    private static void disp(List<List<String>> result) {
        if (result == null || result.size() == 0) {
            System.out.println("EMPTY");
            return;
        }
        System.out.println("[");
        result.forEach(System.out::println);
        System.out.println("]");
    }
}

/*
* 给定一个字符串数组，将字母异位词组合在一起。字母异位词指字母相同，但排列不同的字符串。

示例:

输入: ["eat", "tea", "tan", "ate", "nat", "bat"]
输出:
[
  ["ate","eat","tea"],
  ["nat","tan"],
  ["bat"]
]
说明：

所有输入均为小写字母。
不考虑答案输出的顺序。

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/group-anagrams
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
* */

File: go/year2025/month5/no49/GroupAnagrams.go
package main

import (
	"fmt"
)

func main() {
	strs1 := []string{"eat", "tea", "tan", "ate", "nat", "bat"}
	strs2 := []string{""}
	strs3 := []string{"a"}
	println(fmt.Sprintf("%v", groupAnagrams(strs1)))
	println(fmt.Sprintf("%v", groupAnagrams(strs2)))
	println(fmt.Sprintf("%v", groupAnagrams(strs3)))
}

// HashMap, times: O(nk), space: O(nk)
func groupAnagrams(strs []string) [][]string {
	ans := make(map[string][]string)
	hash := func(s string) (key string) {
		counts := make([]int, 26)
		for _, c := range s {
			counts[c-'a']++
		}
		for _, count := range counts {
			key += fmt.Sprintf("%d#", count)
		}
		return key
	}
	for _, str := range strs {
		hashKey := hash(str)
		if _, ok := ans[hashKey]; !ok {
			ans[hashKey] = make([]string, 0)
		}
		ans[hashKey] = append(ans[hashKey], str)
	}
	result := make([][]string, 0)
	for _, v := range ans {
		result = append(result, v)
	}
	return result
}

/*
给你一个字符串数组，请你将 字母异位词 组合在一起。可以按任意顺序返回结果列表。
字母异位词 是由重新排列源单词的所有字母得到的一个新单词。

示例 1:
输入: strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
输出: [["bat"],["nat","tan"],["ate","eat","tea"]]

示例 2:
输入: strs = [""]
输出: [[""]]

示例 3:
输入: strs = ["a"]
输出: [["a"]]

提示：
1 <= strs.length <= 104
0 <= strs[i].length <= 100
strs[i] 仅包含小写字母
*/
